compiler/src/main/java/dagger/internal/codegen/ComponentDescriptor.java

```diff
@@ -410,9 +412,27 @@ private ComponentDescriptor create(
       for (TypeElement componentDependency : componentDependencyTypes) {
         List<ExecutableElement> dependencyMethods =
             ElementFilter.methodsIn(elements.getAllMembers(componentDependency));
+        HashMultimap<QualifiedMethodNameAndType, Equivalence.Wrapper<TypeMirror>> componentMethodType
+                = HashMultimap.create();
         for (ExecutableElement dependencyMethod : dependencyMethods) {
           if (isComponentContributionMethod(elements, dependencyMethod)) {
-            dependencyMethodIndex.put(dependencyMethod, componentDependency);
+            if (dependencyMethod.getParameters().isEmpty()) {
+              // If a component dependency extends from two interfaces that both contain the same named method, then
```

Comment:
OK, this was my fault.  This wasn't adding up, but I couldn't figure out why.  I think you're implementing a lot of the same logic as `Util.getUnimplementedMethods`.  Let's use that -- the same as we do for implementing a component -- to figure out those details.  If there's a bug, it'll be in that method, if there's not, great.

That'll take care of the naming issues and then we should just be able to take the set of methods that we get back and map them to `Key`s (rather than holding our own types and qualifiers) using `Key.Factory.forComponentMethod()`.

Does that work?